core/arch/x86_64/kernel/thread_64.S

```diff
@@ -15,184 +15,73 @@
-
-/* void thread_rpc_resume(vaddr_t sp, struct thread_smc_args *args,
- *						vaddr_t tmp_stack)
- */
-FUNC thread_rpc_resume , :
-    pushfq
-    pushq %rax
-    pushq %rbx
-    pushq %rcx
-    pushq %rdx
     pushq %rbp
-    pushq %r8
-    pushq %r9
     pushq %r10
     pushq %r11
     pushq %r12
     pushq %r13
     pushq %r14
     pushq %r15
-    pushq %rdi
-    pushq %rsi
-
-    movq %rsp, (%rdx)
-
-    movq %rdi, %rsp
-
-    popq %rdi
-
-    movq (%rsi), %rbx
-    movl %ebx, (%rdi)
-
-    movq 8(%rsi), %rbx
-    movl %ebx, 4(%rdi)
 
-    movq 16(%rsi), %rbx
-    movl %ebx, 8(%rdi)
+    movq %rdi, %r15  //save struct thread_ctx_regs *regs
 
-    movq 24(%rsi), %rbx
-    movl %ebx, 12(%rdi)
+    movq %rsp, %rdi
 
-    movq 32(%rsi), %rbx
-    movl %ebx, 16(%rdi)
+    call thread_set_tmp_sp
 
-    movq 40(%rsi), %rbx
-    movl %ebx, 20(%rdi)
+    movq 8(%r15), %rsp  //switch to thread stack
 
-    popq %rax
-    popq %r15
-    popq %r14
-    popq %r13
-    popq %r12
-    popq %r11
-    popq %r10
-    popq %rbp
-    popq %rbx
-    popq %r9
-    popq %r8
-    popq %rcx
-    popq %rdx
-    popq %rsi
-    popfq
+    pushq (%r15)         //set rip
+    
+    movq 24(%r15), %rdi  //pass parameters for next function call
+    movq 32(%r15), %rsi
+    movq 40(%r15), %rdx
+    movq 48(%r15), %rcx
 
     retq
-END_FUNC thread_rpc_resume
-
-/*void foreign_intr_resume(vaddr_t sp, struct thread_smc_args *args, vaddr_t abt_stack);*/
-FUNC foreign_intr_resume , :
-    pushfq
-    pushq %rax
-    pushq %rbx
-    pushq %rcx
-    pushq %rdx
-    pushq %rbp
-    pushq %r8
-    pushq %r9
-    pushq %r10
-    pushq %r11
-    pushq %r12
-    pushq %r13
-    pushq %r14
-    pushq %r15
-    pushq %rdi
-    pushq %rsi
-
-    movq %rsp, (%rdx)
-
-    movq %rdi, %rsp
+END_FUNC thread_resume
 
-    popq %rdi
-    popq %rax
-    popq %r15
-    popq %r14
-    popq %r13
-    popq %r12
-    popq %r11
-    popq %r10
-    popq %rbp
-    popq %rbx
-    popq %r9
-    popq %r8
-    popq %rcx
-    popq %rdx
-    popq %rsi
-    popfq
+/* void thread_smc_args_retrieve(struct thread_smc_args *args) */
+FUNC thread_smc_args_retrieve , :
+    movq %rsi, (%rdi)
+    movq %rdx, 8(%rdi)
+    movq %rcx, 16(%rdi)
+    movq %r8,  24(%rdi)
 
     retq
-END_FUNC foreign_intr_resume
+END_FUNC thread_smc_args_retrieve
 
 FUNC thread_std_smc_entry , :
-    movq %rsp, %rdi
-
     call __thread_std_smc_entry
-
-    popq %r10
-    popq %r11
-    popq %r12
-    popq %r13
-    popq %r14
-    popq %r15
+    pushq %rax //push return value
 
     cli
     call thread_state_free
 
     call thread_get_tmp_sp
 
-    movq (%rax), %rsp
+    popq %r15 //pop return value
 
-    popq %rsi
+    movq %rax, %rsp
 
-    movq %r10, (%rsi)
-    movq %r11, 8(%rsi)
-    movq %r12, 16(%rsi)
-    movq %r13, 24(%rsi)
-    movq %r14, 32(%rsi)
-    movq %r15, 40(%rsi)
+    movq %r15, %rsi   //prepare parameters for thread_smc_args_retrieve 
+    movq $0,   %rdx
+    movq $0,   %rcx
+    movq $0,   %r8
```

Comment:
Should we use xorq here like "xorq %rdx, %rdx ......" ?

Reply:
updated